logproof/src/linear_relation.rs

```diff
@@ -1404,4 +1409,402 @@ mod test {
+
+                let small_coeff = if coeff > first_coefficient / 2 {
+                    ((coeff as i128) - (first_coefficient as i128)) as i64
+                } else {
+                    coeff as i64
+                };
+
+                *r_i_j = small_coeff;
+            }
+        }
+
+        result
+    }
+
+    /**
+     * Convert a PolynomialArray of small coefficients into a vector of
+     * coefficients. Each outer vector element is one polynomial.
+     */
+    #[allow(unused)]
+    fn convert_to_small_coeffs(
+        coeff_modulus: &[Modulus],
+        poly_array: PolynomialArray,
+    ) -> Vec<Vec<i64>> {
+        convert_to_smallint(coeff_modulus, poly_array)
+            .into_iter()
+            .map(|v| strip_trailing_value(v, 0))
+            .collect()
+    }
+
+    /**
+     * Convert a `PolynomialArray` to a vector of `DensePolynomial`, where all
+     * the coefficients are small.
+     */
+    #[allow(unused)]
+    fn convert_to_polynomial_by_small_coeffs<Q>(
+        coeff_modulus: &[Modulus],
+        poly_array: PolynomialArray,
+    ) -> Vec<DensePolynomial<Q>>
+    where
+        Q: Field,
+    {
+        convert_to_small_coeffs(coeff_modulus, poly_array)
+            .into_iter()
+            .map(|v| make_poly(&v))
+            .collect::<Vec<DensePolynomial<Q>>>()
+    }
+
+    /**
+     * Converts a `PolynomialArray` into a vector of `DensePolynomial`
+     * regardless of the size of the coefficients.
+     */
+    fn convert_to_polynomial<Q, const N: usize>(
+        poly_array: PolynomialArray,
+    ) -> Vec<DensePolynomial<Fp<MontBackend<Q, N>, N>>>
+    where
+        Q: MontConfig<N>,
+    {
+        let chunk_size = poly_array.coeff_modulus_size() as usize;
+
+        let bigint_values = poly_array
+            .as_multiprecision_u64s()
+            .unwrap()
+            .chunks(chunk_size)
+            // SEAL sometimes encodes a multiprecision integer with more limbs
+            // than needed. The trailing limbs can be safely removed since they
+            // are 0.
+            .map(|x| BigInt::<N>(x[0..N].try_into().unwrap()))
+            .collect::<Vec<BigInt<N>>>();
+
+        bigint_values
+            .chunks(poly_array.poly_modulus_degree() as usize)
+            .map(|x| {
+                let leading_zeros_removed = strip_trailing_value(x.to_vec(), BigInt::<N>::zero());
+                DensePolynomial {
+                    coeffs: leading_zeros_removed
+                        .iter()
+                        .map(|y| Fp::<MontBackend<Q, N>, N>::from(*y))
+                        .collect::<Vec<Fp<MontBackend<Q, N>, N>>>(),
+                }
+            })
+            .collect()
+    }
+
+    fn full_knowledge_proof<F, const N: usize>(
+        degree: u64,
+    ) -> LatticeProblem<Fp<MontBackend<F, N>, N>>
+    where
+        F: MontConfig<N>,
+    {
+        let degree = degree;
+
+        let plain_modulus = PlainModulus::raw(1032193).unwrap();
```

Comment:
I think this is too large to N=1024. Maybe we should drop this to like 4096 or some plain modulus that works for N=1024.

Also, might be work moving this to `tests/seal_logproof.rs` and Cargo should pick it up and use it as an E2E test. This file is already enormous.

Reply:
I'll move it.

Turns out `12289` as a plaintext modulus for polynomial modulus degree 1024.

Reply:
Marking as resolved. While batching is still used, a smaller modulus was found that worked for each specific encryption dimension and all tests pass.